test/e2e/apps/statefulset.go

```diff
@@ -630,11 +643,17 @@ var _ = SIGDescribe("StatefulSet", func() {
 
 			By("Verifying that stateful set " + ssName + " was scaled down in reverse order")
 			expectedOrder = []string{ssName + "-2", ssName + "-1", ssName + "-0"}
-			_, err = watch.Until(framework.StatefulSetTimeout, watcher, func(event watch.Event) (bool, error) {
+			ctx, cancel = context.WithTimeout(context.Background(), framework.StatefulSetTimeout)
+			defer cancel()
+			_, err = watchtools.Until(ctx, pl.ResourceVersion, watchFunc, func(event watch.Event) (bool, error) {
+				if event.Type == watch.Error {
```

Comment:
The requirement that a caller now has to handle this is surprising and is going to break people who do the naive `Until` signature converse.n

I don't like that we changed the behavior of Until this dramatically.  I think this PR needs to introduce a new method with the desired behavior and keep the old behavior of Until.

Reply:
Actually that requirement has always been there as that event type was preexisting. 

But I admit we could handle it better. How about we catch status errors in UntilWithoutRetry by type and it returns them as Golang error, ending the loop?